pfl/internal/ops/pytorch_ops.py

```diff
@@ -296,12 +301,42 @@ def __init__(self, seed=None):
 
     def __enter__(self):
         if self._seed is not None:
-            self._saved_state = torch.random.get_rng_state()
+            self._saved_state = {"cpu": torch.random.get_rng_state()}
+            try:
+                self._saved_state["cuda"] = torch.cuda.get_rng_state_all()
+            except Exception as e:
+                logger.info(f"Failed to get cuda rng state: {e}")
+            try:
+                self._saved_state["mps"] = torch.mps.get_rng_state()
+            except Exception as e:
+                logger.info(f"Failed to get mps rng state: {e}")
             torch.random.manual_seed(self._seed)
 
     def __exit__(self, *args):
         if self._seed is not None:
-            torch.random.set_rng_state(self._saved_state)
+            torch.random.set_rng_state(self._saved_state["cpu"])
+            if "cuda" in self._saved_state:
+                torch.cuda.set_rng_state_all(self._saved_state["cuda"])
+            if "mps" in self._saved_state:
+                torch.mps.set_rng_state(self._saved_state["mps"])
+
+
+class Barrier:
```

Comment:
Where is this class used?

Reply:
One example usage is when a horovod script needs to download a very large model/data, this can enforce only 1 process does the download and other processes waits for download to be finished